From d20d39030499ae0d42811184ad4b6acf2f340b87 Mon Sep 17 00:00:00 2001
From: Seal <info@seal.security>
Date: Tue, 6 Jun 2023 16:44:21 +0300
Subject: [PATCH 1/1] Seal Security Hotfix for acorn 5.7.3

This patch fixes:
  GHSA-6chw-6frg-f759

For more information see:
  1. https://seal.security
  2. https://github.com/seal-community
  3. https://github.com/acornjs/acorn/tree/5.7.3
---
 src/regexp.js        | 8 +++++---
 test/tests-regexp.js | 1 +
 2 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/src/regexp.js b/src/regexp.js
index c18ee6d..3fc34fc 100644
--- a/src/regexp.js
+++ b/src/regexp.js
@@ -48,7 +48,8 @@ export class RegExpValidationState {
     if (!this.switchU || c <= 0xD7FF || c >= 0xE000 || i + 1 >= l) {
       return c
     }
-    return (c << 10) + s.charCodeAt(i + 1) - 0x35FDC00
+    const next = s.charCodeAt(i + 1)
+    return next >= 0xDC00 && next <= 0xDFFF ? (c << 10) + next - 0x35FDC00 : c
   }
 
   nextIndex(i) {
@@ -57,8 +58,9 @@ export class RegExpValidationState {
     if (i >= l) {
       return l
     }
-    const c = s.charCodeAt(i)
-    if (!this.switchU || c <= 0xD7FF || c >= 0xE000 || i + 1 >= l) {
+    let c = s.charCodeAt(i), next
+    if (!this.switchU || c <= 0xD7FF || c >= 0xE000 || i + 1 >= l ||
+        (next = s.charCodeAt(i + 1)) < 0xDC00 || next > 0xDFFF) {
       return i + 1
     }
     return i + 2
diff --git a/test/tests-regexp.js b/test/tests-regexp.js
index 6c47194..804e00a 100644
--- a/test/tests-regexp.js
+++ b/test/tests-regexp.js
@@ -1049,6 +1049,7 @@ test("/[\\d][\\12-\\14]{1,}[^\\d]/", {}, { ecmaVersion: 2015 })
 testFail("/[\\d][\\12-\\14]{1,}[^\\d]/u", "Invalid regular expression flag (1:1)", { ecmaVersion: 5 })
 testFail("/[\\d][\\12-\\14]{1,}[^\\d]/u", "Invalid regular expression: /[\\d][\\12-\\14]{1,}[^\\d]/: Invalid class escape (1:1)", { ecmaVersion: 2015 })
 test("/([a ]\\b)*\\b/", {}, { ecmaVersion: 5 })
+test("/[x-*]/u".replace("*", String.fromCharCode(0xd800)), {}, {ecmaVersion: 6})
 
 /*
 // This is test case generator.
-- 
2.39.2 (Apple Git-143)

